File: sample/SampleInProcFunctions.V4/HelperFunctions.cs
// <copyright file="HelperFunctions.cs" company="DarkLoop" author="Arturo Martinez">
//  Copyright (c) DarkLoop. All rights reserved.
// </copyright>

using System;
using System.Security.Claims;
using System.Threading.Tasks;
using Common.Tests;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.Extensions.Logging;

namespace SampleInProcFunctions.V4
{
    public static class HelperFunctions
    {
        [FunctionName("GetTestToken")]
        public static async Task<IActionResult> Run(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = null)] HttpRequest req,
            ILogger log)
        {
            var firstName = "Test";
            var lastName = "User";
            var email = "test.user@domain.com";
            var token = JwtUtils.GenerateJwtToken(new[] {
                new Claim("aud", "api://default"),
                new Claim("iss", "https://localhost/jwt/"),
                new Claim("scp", "user_impersonation"),
                new Claim("tid", Guid.NewGuid().ToString()),
                new Claim("oid", Guid.NewGuid().ToString()),
                new Claim("name", $"{firstName} {lastName}"),
                new Claim(ClaimTypes.Name, email),
                new Claim(ClaimTypes.Upn, email),
                new Claim(ClaimTypes.Email, email),
                new Claim(ClaimTypes.GivenName, firstName),
                new Claim(ClaimTypes.Surname, lastName),
                new Claim("role", "Just a user"),
                new Claim("role", "another user"),
            });

            return await Task.FromResult<IActionResult>(new OkObjectResult(token));
        }
    }
}

File: sample/SampleIsolatedFunctionsFSharp.V4/HelperFunctions.fs
namespace SampleInProcFunctions.V4

open System.Security.Claims
open Common.Tests
open Microsoft.AspNetCore.Http
open Microsoft.AspNetCore.Mvc
open Microsoft.Azure.Functions.Worker
open Microsoft.Extensions.Logging
open System

type HelperFunctions() =    
    
    [<Function("GetTestToken")>]
    member _.Run(
            [<HttpTrigger("get", Route = null)>]
            req: HttpRequest,
            log: ILogger) =
        task {

                let firstName = "Test"
                let lastName = "User"
                let email = "test.user@domain.com"
                let token = JwtUtils.GenerateJwtToken(
                    [
                        new Claim("aud", "api://default")
                        new Claim("iss", "https://localhost/jwt/")
                        new Claim("scp", "user_impersonation")
                        new Claim("tid", Guid.NewGuid().ToString())
                        new Claim("oid", Guid.NewGuid().ToString())
                        new Claim("name", $"{firstName} {lastName}")
                        new Claim(ClaimTypes.Name, email)
                        new Claim(ClaimTypes.Upn, email)
                        new Claim(ClaimTypes.Email, email)
                        new Claim(ClaimTypes.GivenName, firstName)
                        new Claim(ClaimTypes.Surname, lastName)
                        new Claim("role", "Just a user")
                        new Claim("role", "admin")
                    ])

            return OkObjectResult(token)
        }


